src/gadgets/sumcheck.rs

```diff
@@ -0,0 +1,336 @@
+    let power_aux = power.mul(cs.namespace(|| format!("mul2 {}", i)), &r)?;
+    // r1cs constr: power_aux = power * r
+    cs.enforce(
+      || format!("power_aux = power * r {}", i),
+      |lc| lc + power.get_variable(),
+      |lc| lc + r.get_variable(),
+      |lc| lc + power_aux.get_variable(),
+    );
+    power = power_aux;
+  }
+  let allocated_eval: AllocatedNum<Scalar> =
+    AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+
+  Ok(allocated_eval)
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::bellperson::{
+    r1cs::{NovaShape, NovaWitness},
+    {shape_cs::ShapeCS, solver::SatisfyingAssignment},
+  };
+  use crate::spartan::sumcheck::UniPoly;
+  use pasta_curves::{arithmetic::CurveAffine, pallas, pallas::Scalar, vesta};
+
+  use crate::traits::{Group, TranscriptEngineTrait, TranscriptReprTrait};
+  type PastaG1 = pasta_curves::pallas::Point;
+  use crate::spartan::polynomial::MultilinearPolynomial;
+  use crate::spartan::sumcheck::SumcheckProof;
+
+  fn synthetize_uni_evaluate<Scalar, CS>(
+    mut cs: CS,
+    coeffs: Vec<Scalar>,
+    r: Scalar,
+  ) -> AllocatedNum<Scalar>
+  where
+    Scalar: PrimeField + PrimeFieldBits,
+    CS: ConstraintSystem<Scalar>,
+  {
+    // prepare inputs
+    let r_var = AllocatedNum::<Scalar>::alloc(cs.namespace(|| "alloc r"), || Ok(r)).unwrap();
+    let mut coeffs_var: Vec<AllocatedNum<Scalar>> = Vec::new();
+    for (i, coeff) in coeffs.iter().enumerate() {
+      coeffs_var.push(
+        AllocatedNum::<Scalar>::alloc(cs.namespace(|| format!("alloc coeffs {}", i)), || {
+          Ok(*coeff)
+        })
+        .unwrap(),
+      );
+    }
+
+    // define inputs
+    // let _ = r_var.inputize(cs.namespace(|| "Input r"));
+    // for (i, coeff) in coeffs_var.iter().enumerate() {
+    //   let _ = coeff.inputize(cs.namespace(|| format!("Input coeffs {}", i)));
+    // }
+
+    // evaluate in-circuit (synthetize)
+    let res = uni_evaluate(cs.namespace(|| "uni_evaluate"), coeffs_var, r_var).unwrap();
+    // let _ = res.inputize(cs.namespace(|| "Output res"));
+
+    res
+  }
+
+  #[test]
+  fn test_uni_evaluate() {
+    let evals = vec![
+      Scalar::from(1_u64),
+      Scalar::from(2_u64),
+      Scalar::from(3_u64),
+    ];
+    let p = UniPoly::<pallas::Point>::from_evals(&evals);
+    let r = Scalar::from(5_u64);
+
+    // let's test it against the CS
+    let mut cs: ShapeCS<pallas::Point> = ShapeCS::new();
+    let _ = synthetize_uni_evaluate(&mut cs, p.coeffs.clone(), r);
+    println!("Number of constraints: {}", cs.num_constraints());
+
+    let (shape, ck) = cs.r1cs_shape();
+
+    let mut cs: SatisfyingAssignment<pallas::Point> = SatisfyingAssignment::new();
+    let res = synthetize_uni_evaluate(&mut cs, p.coeffs.clone(), r);
+    assert_eq!(res.get_value().unwrap(), p.evaluate(&r));
+
+    let (inst, witness) = cs.r1cs_instance_and_witness(&shape, &ck).unwrap();
+    assert!(shape.is_sat(&ck, &inst, &witness).is_ok());
```

Comment:
How can we have some basic negative tests here, to make sure that not everything is satisfied?
Can we make a shape that is obviously not satisfied and then check that it is indeed not satisfied?

Reply:
The `uni_evaluate` gadget does not check that two inputs satisfy the checks, it internally computes the evaluation of the univariate polynomial and outputs the result, while internally adding constraints for the operations of the evaluation.
But there is no 'wrong inputs' for that, what we can check is that the output of the gadget is the expected result of the evaluation (which is what the test does).

Is like a gadget that we can name "add" that outputs the result of `a + b`, we can not input 'wrong' values that will not satisfy the shape, as it is just computing a+b and adding a constraint that ensures that the output is equal to a+b, but to test that gadget we can instanciate it and assert that the outputted value is the expected one.